src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs

```diff
@@ -0,0 +1,2164 @@
+                    throw new ArgumentOutOfRangeException(
+                        "value",
+                        value,
+                        string.Format(SR.net_http_value_must_be_greater_than, 0));
+                }
+
+                CheckDisposedOrStarted();
+                _maxResponseHeadersLength = value;
+            }
+        }
+
+        public int MaxResponseDrainSize
+        {
+            get
+            {
+                return _maxResponseDrainSize;
+            }
+
+            set
+            {
+                if (value <= 0)
+                {
+                    throw new ArgumentOutOfRangeException(
+                        "value",
+                        value,
+                        string.Format(SR.net_http_value_must_be_greater_than, 0));
+                }
+
+                CheckDisposedOrStarted();
+                _maxResponseDrainSize = value;
+            }
+        }
+        #endregion
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing && !_disposed)
+            {
+                _disposed = true;
+
+                if (_sessionHandle != null)
+                {
+                    _sessionHandle.DangerousRelease();
+                    SafeInternetHandle.DisposeAndClearHandle(ref _sessionHandle);
+                }
+            }
+
+            base.Dispose(disposing);
+        }
+
+#if HTTP_DLL
+        protected internal override Task<HttpResponseMessage> SendAsync(
+            HttpRequestMessage request,
+            CancellationToken cancellationToken)
+#else
+        protected override Task<HttpResponseMessage> SendAsync(
+            HttpRequestMessage request,
+            CancellationToken cancellationToken)
+#endif
+        {
+            if (request == null)
+            {
+                throw new ArgumentNullException("request", SR.net_http_handler_norequest);
+            }
+
+            // Check for invalid combinations of properties.
+            if (_proxy != null && _windowsProxyUsePolicy != WindowsProxyUsePolicy.UseCustomProxy)
+            {
+                throw new InvalidOperationException(SR.net_http_invalid_proxyusepolicy);
+            }
+
+            if (_windowsProxyUsePolicy == WindowsProxyUsePolicy.UseCustomProxy && _proxy == null)
+            {
+                throw new InvalidOperationException(SR.net_http_invalid_proxy);
+            }
+
+            if (_cookieUsePolicy == CookieUsePolicy.UseSpecifiedCookieContainer &&
+                _cookieContainer == null)
+            {
+                throw new InvalidOperationException(SR.net_http_invalid_cookiecontainer);
+            }
+
+            CheckDisposed();
+
+            SetOperationStarted();
+
+            TaskCompletionSource<HttpResponseMessage> tcs = new TaskCompletionSource<HttpResponseMessage>();
+
+            // Create RequestState object and save current values of handler settings.
+            RequestState state = new RequestState();
```

Comment:
As we're looking at perf later, there are a variety of improvements that can be made here.  For example, it looks like RequestState could derive from TaskCompletionSource instead of having a field of one, and it could have a field to store ```this```.  That has a few benefits.  First, it shrinks the size of the combination of RequestSize + TaskCompletionSource by several fields (state.Tcs, the extra object header, etc.)  Second, it means that the state passed to StartNew already contains the ```this``` reference necessary to call the instance StartRequest method... that lets you rewrite the call to use a lambda like ```s => ((RequestState)s).thisRef.StartRequest()```, which in turn lets the compiler cache the delegate and avoid the delegate allocation this currently incurs.